src/ImageSearchModal.jsx

```diff
@@ -0,0 +1,169 @@
+        this._isMounted = true;
+    }
+
+    componentWillUnmount() {
+        this._isMounted = false;
+
+        if (this.activeConnection)
+            this.activeConnection.close();
+    }
+
+    onDownloadClicked() {
+        let selectedImageName = this.state.imageList[this.state.selected].name;
+
+        this.props.close();
+        this.props.downloadImage(selectedImageName);
+    }
+
+    onItemSelected(key) {
+        this.setState({ selected: key });
+    }
+
+    onSearchTriggered(forceSearch) {
+        // When search re-triggers close any existing active connection
+        if (this.activeConnection)
+            this.activeConnection.close();
+        this.setState({ searchFinished: false });
+
+        // Do not call the SearchImage API if the input string  is not at least 2 chars,
+        // unless Enter is pressed, which should force start the search.
+        // The comparison was done considering the fact that we miss always one letter due to delayed setState
+        if (this.state.imageIdentifier.length < 2 && !forceSearch)
+            return;
+
+        this.setState({ searchInProgress: true });
+
+        varlink.connect(utils.PODMAN_ADDRESS)
+                .then(connection => {
+                    this.activeConnection = connection;
+
+                    connection.call("io.podman.SearchImage", { name: this.state.imageIdentifier })
+                            .then(reply => {
+                                if (this._isMounted)
+                                    this.setState({ imageList: reply.images || [], searchInProgress: false, searchFinished: true });
+                            })
+                            .catch(ex => {
+                                // We expect new searches to close the connection for ongoing searches
+                                if (ex.error == 'ConnectionClosed')
+                                    return;
+
+                                if (this._isMounted) {
+                                    this.setState({
+                                        searchInProgress: false,
+                                        dialogError: _("Failed to search for new images"),
+                                        dialogErrorDetail: cockpit.format("$0: $1", ex.error, ex.parameters && ex.parameters.reason)
+                                    });
+                                }
+                            });
+                });
+    }
+
+    onValueChanged(key, value) {
+        if (key == 'imageIdentifier')
+            this.setState({ [key]: value.trim() });
+        else
+            this.setState({ [key]: value });
+    }
+
+    onKeyPress(e) {
+        if (e.key != ' ') { // Space should not trigger search
+            // Clears the previously set timer.
+            clearTimeout(this.typingTimeout);
+
+            // Reset the timer, to make the http call after 250MS
+            this.typingTimeout = setTimeout(this.onSearchTriggered(e.key == 'Enter'), 250);
```

Comment:
This is calling `onSearchTriggered` immediately instead of after 250ms. That's why @martinpitt and @garrett were seeing searches after the first key press.

Also, it leads to a CSP error:  `setTimeout` tries to call the return value of `onSearchTriggered` and since that's not a function, calls `eval()` on it. Yay JavaScript.

This patch below makes it work. I have the extra variable `forceSearch` because we can't access `event` after the handler has run. React reuses the event object and complains if we do.

```diff
     onKeyPress(e) {
         if (e.key != ' ') { // Space should not trigger search
+            let forceSearch = e.key == 'Enter';
+
             // Clears the previously set timer.
             clearTimeout(this.typingTimeout);
 
             // Reset the timer, to make the http call after 250MS
-            this.typingTimeout = setTimeout(this.onSearchTriggered(e.key == 'Enter'), 250);
+            this.typingTimeout = setTimeout(() => this.onSearchTriggered(forceSearch), 250);
```

Reply:
Ups, thanks for noticing it @larskarlitski!